src/powershell/Program.cs

```diff
@@ -19,7 +20,232 @@ public sealed class ManagedPSEntry
+            for (int i = 0; i < args.Length; i++)
+            {
+                string arg = args[i];
+
+                // Must look like '-<name>'
+                if (arg == null || arg.Length < 2 || arg[0] != '-')
+                { 
+                    continue;
+                }
+
+                // Check for "-Login" or some prefix thereof
+                if (IsParam(arg, "login", "LOGIN"))
+                {
+                    loginIndex = i;
+                    return true;
+                }
+
+                // After -File and -Command, all parameters are passed
+                // to the invoked file or command, so we can stop looking.
+                if (IsParam(arg, "file", "FILE")
+                    || IsParam(arg, "command", "COMMAND"))
+                {
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines if a given parameter is the one we're looking for.
+        /// Assumes any prefix determines that parameter (true for -l, -c and -f).
+        /// </summary>
+        /// <param name="arg">The argument to check.</param>
+        /// <param name="paramToCheck">The lowercase name of the parameter to check.</param>
+        /// <param name="paramToCheckUpper">The uppercase name of the parameter to check.</param>
+        /// <returns></returns>
+        private static bool IsParam(
+            string arg,
+            string paramToCheck,
+            string paramToCheckUpper)
+        {
+            // Quick fail if the argument is longer than the parameter
+            if (arg.Length > paramToCheck.Length + 1)
+            {
+                return false;
+            }
+
+            // Check arg chars in order and allow prefixes
+            for (int i = 1; i < arg.Length; i++)
+            {
+                if (arg[i] != paramToCheck[i-1]
+                    && arg[i] != paramToCheckUpper[i-1])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Create the exec call to /bin/{ba}sh -l -c 'pwsh "$@"' and run it.
+        /// </summary>
+        /// <param name="args">The argument vector passed to pwsh.</param>
+        /// <param name="loginArgIndex">The index of -Login in the argument vector.</param>
+        /// <returns>
+        /// The exit code of exec if it fails.
+        /// If exec succeeds, this process is overwritten so we never actually return.
+        /// </returns>
+        private static int ExecPwshLogin(string[] args, int loginArgIndex)
+        {
+            // We need the path to the current pwsh executable
+            string pwshPath = System.Diagnostics.Process.GetCurrentProcess().MainModule.FileName;
+
+            // Create input for /bin/sh that execs pwsh
+            int quotedPwshPathLength = GetQuotedPathLength(pwshPath);
```

Comment:
Does `pwshPath` actually get quoted?  I checked on Windows and macOS, and even with whitespace there's no quoting returned.

Reply:
Looks like this on my machine:

<img width="1009" alt="Screen Shot 2019-07-08 at 16 39 11" src="https://user-images.githubusercontent.com/7009879/60849306-fcd17180-a19e-11e9-942f-6758546b829c.png">

The invocation looks like:

```
"exec '/Users/rjmholt/Documents/Dev/Microsoft/PowerShell/src/powershell-unix/bin/Debug/netcoreapp3.0/osx-x64/publish/pwsh' \"$@\""
```

How did you check in your case?

Also, this code path doesn't exist on Windows.

Reply:
Oh I see you mean `pwshPath`; no that's what .NET gives us. That path is quoted inside `CreatePwshInvocation`, so no intermediate strings are created.

Reply:
It seems that at this point `pwshPath` doesn't have any quotes, so the call to `GetQuotedPathLength()` is unnecessary.

Reply:
The path may have whitespace in it, in which case we need to put quotes around the whole thing for `exec`.

Imagine PowerShell is installed to `/Users/Rob Holt/apps/PowerShell/pwsh`, then not quoting will produce:

```sh
exec /Users/Rob Holt/apps/PowerShell/pwsh "$@"
```

In that case (which we support, because of binplace installation), exec will fail. So instead we quote it:

```sh
exec '/Users/Rob Holt/apps/PowerShell/pwsh' "$@"
```

Reply:
I also doubt that this is necessary as we pass the parameters in execv "as is" - no parsing.

Reply:
This is a single parameter to execv. It looks like this:

```
execv("/bin/sh", "-l", "-c", "exec '/Path/to/pwsh' \"$@\"", <args>, <to>, <pwsh>)
```

(It's slightly more complicated than that, but this illustrates the required logic).

We exec twice, once with the syscall to get /bin/sh and once in /bin/sh with the builtin. The second case is a short shell command, and in that the path to pwsh needs escaping so that the `exec` *builtin* hits the right executable.

Reply:
If we will consider only login shell scenario perhaps we could simplify preparing of the  arguments.